### ISSUE TITLE
Feature Request: Allow buffer_size to be configured in UDP file options

### ISSUE THREAD

**Issue by kqyang** [google/shaka-packager#411]
Sometimes we saw small GAPs as in https://github.com/google/shaka-packager/issues/390. One possible cause is buffer overrun caused UDP packet loss.

Having a configurable buffer_size will allow `receive buffer size` to be increased to mitigate packet loss due to `receive buffer overrun`, which is a common cause for packet loss.

**Comment by jakubvojacek** [comment]
Hello,

got any ETA on this feature? In before you implement it, is the current receiver buffer size somewhere hardcoded in the shaka packager source code? Can we manually increase it and re-compile ourselves to be able to test this? I tried looking into the code but could not find it.  Would be good to know whether multicast will work or we will have to stick with pipes.

Thank you

**Comment by kqyang** [comment]
@jakubvojacek It has been merged. Documentation is also updated: https://google.github.io/shaka-packager/html/tutorials/live.html#udp-file-options. Let us know if it works for you.

**Comment by jakubvojacek** [comment]
@kqyang doing tests since morning, too early to say it works.

One thing that I noticed though - when using ` --io_cache_size 200000000`, the CPU consumed by packager is like 2-3x more than before, is this normal? Had to remove this flag now because it was causing too much load. Is the number 200000000 too high?

**Comment by kqyang** [comment]
That is probably a bit high. I am not very familiar with how Linux kernel works. I guess it causes physical memory to be used up and result in a lot of page swapping, which increases CPU usage.

**Comment by jakubvojacek** [comment]
thank you, testing with 50mb now, the CPU usage is now around 10-15% higher which is doable. Will post results once I got some

**Comment by jakubvojacek** [comment]
@kqyang  I had it running for two days on around 100 channels now, for channels that there are no issues with the signal / input, the results are perfect, no gaps. 

Thank you again!

**Comment by kqyang** [comment]
@jakubvojacek Good to know! Thanks for the update. Btw, what is your setting of `--io_cache_size` and `buffer_size` option respectively? Does it work if you keep `--io_cache_size` as is and just modify `buffer_size` only?

**Comment by jakubvojacek** [comment]
I am setting both of these at this moment, `--io_cache_size` is at `50000000` and `buffer_size` at `150000000`. I think those are still way too high and I will be trying to make them lower (as the packager consumes more CPU with these higher values) but right now, I am just happy that it plays. I was consulting it with a colleague today and we believe we could go as low as 20% of what we're setting now.

What are the default values for `--io_cache_size` and `buffer_size`, please?

**Comment by kqyang** [comment]
The default value for `--io_cache_size` is 2MB and the default value for `buffer_size` varies by platform. It is ~200KB on my system (you can check the value using `sysctl net.core.rmem_default`).

**Comment by rkuroiwa** [comment]
Which makes me think it might be a good idea to add another verbose logging, the max size of the data in IoCache. It would be nice stats for users.

**Comment by kqyang** [comment]
@rkuroiwa Good idea. We should print the value in the io cache buffer full verbose logging: https://github.com/google/shaka-packager/commit/7bc90004e0214840664eb49ee8df5e273c24718c. I'll prepare a CL for that.
